core/store/models/log_events.go

```diff
@@ -85,55 +86,75 @@ var topicFactoryMap = map[common.Hash]logRequestParser{
 	RunLogTopic20190207withoutIndexes:         parseRunLog20190207withoutIndexes{},
 }
 
-// TopicFiltersForRunLog generates the two variations of RunLog IDs that could
-// possibly be entered on a RunLog or a ServiceAgreementExecutionLog. There is the ID,
-// hex encoded and the ID zero padded.
-func TopicFiltersForRunLog(logTopics []common.Hash, jobID *ID) [][]common.Hash {
-	return [][]common.Hash{logTopics, {IDToTopic(jobID), IDToHexTopic(jobID)}}
+// LogBasedChainlinkJobInitiators are initiators which kick off a log-specified
+// chainlink job when an appropriate ethereum log is received.
+// (InitiatorFluxMonitor kicks off a job when a log is emitted, but the JobID is
+// not taken from the log)
+var LogBasedChainlinkJobInitiators = []string{InitiatorRunLog, InitiatorEthLog,
+	InitiatorServiceAgreementExecutionLog}
+
+// topicsForInitiatorsWhichRequireJobSpecTopic are the log topics which kick off
+// a user job with the given type of initiator. If chainlink has any jobs with
+// these initiators, it subscribes on startup to logs which match both these
+// topics and some representation of the job spec ID.
+var TopicsForInitiatorsWhichRequireJobSpecIDTopic = map[string][]common.Hash{
+	InitiatorRunLog: {RunLogTopic20190207withoutIndexes,
+		RunLogTopic20190123withFullfillmentParams, RunLogTopic0original},
+	InitiatorServiceAgreementExecutionLog: {ServiceAgreementExecutionLogTopic},
+}
+
+// initiationRequiresJobSpecId is true if jobs initiated by the given
+// initiatiatorType require that their initiating logs match their JobSpecIDs.
+func initiationRequiresJobSpecID(initiatorType string) bool {
+	_, ok := TopicsForInitiatorsWhichRequireJobSpecIDTopic[initiatorType]
+	return ok
+}
+
+// jobSpecIDTopics lists the ways jsID could be represented as a log topic. This
+// allows log subscriptions to respond to all possible representations.
+func JobSpecIDTopics(jsID *ID) []common.Hash {
+	return []common.Hash{
+		// The job to be initiated can be encoded in a log topic in two ways:
+		IDToTopic(jsID),    // 16 full-range bytes, left padded to 32 bytes,
+		IDToHexTopic(jsID), // 32 ASCII hex chars representing the 16 bytes
+	}
 }
 
 // FilterQueryFactory returns the ethereum FilterQuery for this initiator.
-func FilterQueryFactory(i Initiator, from *big.Int) (ethereum.FilterQuery, error) {
-	q := ethereum.FilterQuery{
-		FromBlock: from,
-		Addresses: utils.WithoutZeroAddresses([]common.Address{i.Address}),
-	}
+func FilterQueryFactory(i Initiator, from *big.Int) (q ethereum.FilterQuery, err error) {
+	q.FromBlock = from
+	q.Addresses = utils.WithoutZeroAddresses([]common.Address{i.Address})
 
-	switch i.Type {
-	case InitiatorEthLog:
+	switch {
```

Comment:
This code would be clearer if the safe path returned a new literal FilterQuery and the case blocks just set a `FromBlock` variable. IMHO.

Reply:
You're right, that's much nicer. Thanks.